design/metadata-handling.md

```diff
@@ -0,0 +1,668 @@
+      - vlan:
+          id: "vlan1"
+          mtu: 1500
+          macAddress:
+            string: "YY:YY:YY:YY:YY:YY"
+          vlanId: 1
+          vlanLink: bond0
+    networks:
+      - ipv4Dhcp:
+          id: "provisioning"
+          link: "bond0"
+
+      - ipv4:
+          id: "Baremetal"
+          link: "vlan1"
+          ipAddress:
+            start: "192.168.0.10"
+            end: "192.168.0.100"
+            subnet: "192.168.0.0/24"
+            step: 1
+          netmask: 24
+          routes:
+            - network: "0.0.0.0"
+              netmask: 0
+              gateway: "192.168.0.1"
+              services:
+                - type: "dns"
+                  address: "8.8.4.4"
+      - ipv6Dhcp:
+          id: "provisioning6"
+          link: "bond0"
+      - ipv6Slaac:
+          id: "provisioning6slaac"
+          link: "bond0"
+      - ipv6:
+          id: "Baremetal6"
+          link: "vlan1"
+          ipAddress:
+            start: "2001:0db8:85a3::8a2e:0370:a"
+            end: "2001:0db8:85a3::8a2e:0370:fff0"
+            subnet: "2001:0db8:85a3::8a2e:0370:0/64"
+            step: 10
+          netmask: 64
+          routes:
+            - network: "0::0"
+              netmask: 0
+              gateway: "2001:0db8:85a3::8a2e:0370:1"
+              services:
+                - dns: "2001:4860:4860::8844"
+    services:
+      - dns: "8.8.8.8"
+      - dns: "2001:4860:4860::8888"
+```
+
+This object will be reconciled by its own controller. When reconciled,
+the controller will add a label pointing to the Metal3Cluster that has nodes
+linking to this object. The spec contains a `metaData` and a `networkData` field
+that contain a template of the values that will be rendered for all nodes.
+
+The `metaData` field will be rendered into a map of strings in yaml format,
+while `networkData` will be rendered into a json string that fulfills the
+requirements of
+[Nova network_data.json](https://docs.openstack.org/nova/latest/user/metadata.html#openstack-format-metadata).
+The format definition can be found
+[here](https://docs.openstack.org/nova/latest/_downloads/9119ca7ac90aa2990e762c08baea3a36/network_data.json).
+
+#### Metadata Specifications
+
+The `metaData` field contains a list of items that will render data in different
+ways. The following objects are available:
+
+- **key**: This is compulsory and is the key used for the rendered value in
+  the metadata map.
+- **string**: renders the given string as value in the metadata
+- **objectName** : renders the name of the object that matches the type given in
+  the `object` field.
+- **index**: renders the index of the current object, with the offset from the
+  `offset` field and using the step from the `step` field.
+- **ip**: renders an ip address based on the index, based on the `start` value
+  if given or using `subnet` to calculate the start value, and checking that
+  the rendered value is not over the `end` value. The increment is the `step`
+  value.
```

Comment:
What happens if the computed value goes out of bounds?

Reply:
If the computed value goes out of bounds, the error status will be set
  with the error in the error message.